### ISSUE TITLE
xdg-open zombies

### ISSUE THREAD

**Issue by pdobsan** [jarun/nnn#449]
Opening any files via `xdg-open` leaves the `xdg-open` behind as a zombie process after the launched process finished. The `xdg-open` zombies are cleaned up only after `nnn` exited.

For example, here is the result after clicking on a few files (.pdf, .mp4 etc.), viewing finshed, `nnn` still running.

> `% ps x | grep xdg-op`
  `90044 ?        Zs     0:00 [xdg-open] <defunct>`
  `90146 ?        Zs     0:00 [xdg-open] <defunct>`
  `90247 ?        Zs     0:00 [xdg-open] <defunct>`

#### Environment details (Put `x` in the checkbox along with the information)

[x] Operating System: archlinux
[x] Desktop Environment: x11
[x] Terminal Emulator: any (urxvt, termite, ...)
[x] Shell: bash, zsh
[ ] Custom desktop opener (if applicable):
[x] Issue exists on `nnn` master

**Comment by jarun** [comment]
This was reported in #227 and I finally closed it without a fix because of the issues seen if we try to address it.

I do not see the problem if I use nuke. I am closing this as a duplicate of #227.

@KlzXS and @annagrram do share if you have some insights on #227.

**Comment by KlzXS** [comment]
As I understand detaching it would solve the problem, why do we not do that?

**Comment by jarun** [comment]
The details are in #227. If we wait it doesn't become a zombie.

However, when we wait and the terminal is closed directly, all spawned apps are also killed.

I think the best solution is to make 0 (init) the parent of the spawned process (like xdg-open). Can you try that?

**Comment by KlzXS** [comment]
> I think the best solution is to make 1 (init) the parent of the spawned process (like xdg-open). Can you try that?

That's what I was going after. You just do a double `fork()`. fork the parent, fork the child, `exit()` the child. The grandchild is now an orphan under init. And perhaps a `setsid()` to make it more independent I guess.

**Comment by jarun** [comment]
We do `setsid()`. Give it a try.

**Comment by jarun** [comment]
Fixed at commit e2d37632eaa9599c344851cfeb9b77851cd97281.

**Comment by pdobsan** [comment]
I can confirm it works fine. An alternative solution to the double `fork` could have been using `sigaction (2)` with `SA_NOCLDWAIT` -- a bit late advice, I know.

Thank you for the quick fix.

**Comment by jarun** [comment]
@pdobsan we would love to avoid the double fork. And it's never too late for that. Can you please raise a PR?

**Comment by jarun** [comment]
On second thoughts, I think that would have issues with spawning cli mode children, when we need to wait.

**Comment by pdobsan** [comment]
You probably are right. Also, to raise a proper PR I would need to completely understand the whole program. Anyway, here is a skeleton `xfork()` which seems to work. You might need to do more work in the arms of the switch statement.

``` C
static pid_t xfork(uchar flag)
{
    struct sigaction act;
    pid_t p;

    memset(&act, 0, sizeof(act));
    act.sa_handler = SIG_IGN;
    act.sa_flags |= SA_RESTART | SA_NOCLDSTOP; // no SIGCHLD if child stops
    if (flag & F_NOWAIT) {
        if ((flag & F_NOWAIT) && sigaction(SIGCHLD, &act, 0) == -1) {
            perror("sigaction");
            exit(-1);
        }
    }

    p = fork();
    switch (p) {
        case -1:
            perror("fork");
            return p;
        case 0:         // child
            setsid();
            return p;
        default:        // parent
            return p;
    }
}
```

**Comment by jarun** [comment]
Did you test with a case where we need to wait? For example, by having `-e` and opening a text file in vim as $EDITOR?

**Comment by pdobsan** [comment]
> Did you test with a case where we need to wait? For example, by having `-e` and opening a text file in vim as $EDITOR?

No, I did not. Trying now ... it does not work.

**Comment by jarun** [comment]
It seemed so... the current solution works in all cases.
